software/bsg_manycore_lib/bsg_mcs_mutex.S

```diff
@@ -0,0 +1,80 @@
+// MCS mutex
+// Author: Max
+//
+// This is an implementation of the MCS mutex described by Mellor-Crummey and Scott in their 1991 paper
+// “Algorithms for Scalable Synchronization on Shared-Memory Multiprocessors”
+//
+// This is a spinlock mutex, but unlike a simple spinlock in which all threads update and spin on
+// a single memory location, the MCS lock builds a linked-list of memory locations local to each core.
+//
+// Cores atomically append their local memory region to the global list using an unconditional
+// amoswap operation. They then spin on their local memories for a predecessor in the queue
+// to notify them that they now hold the lock.
+//
+// Once a core has completed its critical region, it checks for a successor and updates releases the lock to them.
+//
+// The advantages of this mutex over a simple spin lock on the manycore are two fold:
+//
+// (1) It greatly reduces the number of memory reqeusts on the network and it mitigates the extent to which
+// a single memory bank becomes a hot-spot. The number of requests issued to a memory bank containing the
+// lock object is linear with the number of times an acquire operation is execution.
+//
+// (2) The lock approximates a FIFO-ish structure, which improves fairness. A simple spinlock on the manycore
+// will favor threads topologically closer to the memory bank in which the lock resides and can lead to
+// starvation of the other cores.
+//
+// This lock is by no means perfect. For locks with low contention, a simple spinlock may result in better performance.
+
+        .text
+        .globl bsg_mcs_mutex_acquire
+bsg_mcs_mutex_acquire:
+        // a0 = mtx         : bsg_mcs_mutex_t*
```

Comment:
- definition and naming of input parameters?  
- probably a command to clarify that the global pointer would preferably a tile group global pointer, unless the synchronization is being done between tile groups; this would preserve virtualization of tile groups

Reply:
indication that a0 must be in DRAM address space?

Reply:
a1/a2 must be in the tile's local DMEM?

Reply:
see the header bsg_mcs_mutex.h, I can add them here too.

Reply:
I missed the dram address space bit though

Reply:
you can have a comment that says 

// refer to bsg_mcs_mutex.h for detailed description of usage